clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -1981,4 +1992,145 @@ void handleFailure(Throwable throwable) {
+                                    new MemberAssignment(assignment.partitions()));
+                            consumers.add(memberDescription);
+                        }
+                        final String protocol = groupMetadata.protocol();
+                        final ConsumerGroupDescription consumerGroupDescription =
+                            new ConsumerGroupDescription(groupId, protocolType.isEmpty(), consumers, protocol);
+                        future.complete(consumerGroupDescription);
+                    }
+                }
+            }
+
+            @Override
+            boolean handleUnsupportedVersionException(UnsupportedVersionException exception) {
+                return false;
+            }
+
+            @Override
+            void handleFailure(Throwable throwable) {
+                completeAllExceptionally(consumerGroupFutures.values(), throwable);
+            }
+        }, now);
+        return new DescribeConsumerGroupsResult(new HashMap<String, KafkaFuture<ConsumerGroupDescription>>(consumerGroupFutures));
+    }
+
+    @Override
+    public ListConsumerGroupsResult listConsumerGroups(ListConsumerGroupsOptions options) {
+        final KafkaFutureImpl<Map<String, ConsumerGroupListing>> groupListingFuture = new KafkaFutureImpl<>();
+        final long now = time.milliseconds();
+        runnable.call(new Call("listConsumerGroups", calcDeadlineMs(now, options.timeoutMs()),
+            new LeastLoadedNodeProvider()) {
+
+            @Override
+            AbstractRequest.Builder createRequest(int timeoutMs) {
+                return new ListGroupsRequest.Builder();
+            }
+
+            @Override
+            void handleResponse(AbstractResponse abstractResponse) {
+                final ListGroupsResponse response = (ListGroupsResponse) abstractResponse;
+                final Map<String, ConsumerGroupListing> groupsListing = new HashMap<>();
+                for (ListGroupsResponse.Group group : response.groups()) {
+                    if (group.protocolType().equals(ConsumerProtocol.PROTOCOL_TYPE) || group.protocolType().isEmpty()) {
+                        final String groupId = group.groupId();
+                        final String protocolType = group.protocolType();
+                        final ConsumerGroupListing groupListing = new ConsumerGroupListing(protocolType.isEmpty());
+                        groupsListing.put(groupId, groupListing);
+                    }
+                }
+                groupListingFuture.complete(groupsListing);
+            }
+
+            @Override
+            void handleFailure(Throwable throwable) {
+                groupListingFuture.completeExceptionally(throwable);
+            }
+        }, now);
+        return new ListConsumerGroupsResult(groupListingFuture);
+    }
+
+    @Override
+    public ListConsumerGroupOffsetsResult listConsumerGroupOffsets(final String groupId, final ListConsumerGroupOffsetsOptions options) {
```

Comment:
Like DescribeGroups, we need to find the coordinator for the group to send the OffsetFetch request to.